tensorflow/compiler/xla/service/slice_sinker.cc

```diff
@@ -0,0 +1,254 @@
+// it should be changed.
+// This routine assumes that we will generate a new elementwise operation using
+// the "slice_sources" as operands. For the following example, we only need the
+// result of add([0:9] slice of p) while we generate add(p).
+//
+// p = f32[10] parameter(0)
+// a = f32[8] slice(p), slice=[0:8]
+// aa = add(a, a)
+// b = f32[7] slice(p), slice=[2:9]
+// bb = add(b, b)
+// (there are 15 scalar add operations, but only 10 scalar add operations when
+// add(p) )
+bool ShouldReplace(const std::vector<HloInstruction*>& operands,
+    const std::vector<HloInstruction*>& users) {
+  int64 sum = 0;
+  // Sums the total element number of the peer operations.
+  for (HloInstruction* user : users) {
+    sum += ShapeUtil::ElementsIn(user->shape());
+  }
+  // Operand and user have the same element number in shape because of
+  // elementwise operation, so the elements in operands[0] is the same as in
+  // new user if new user is generated.
+  // Compares the total elements in all peer operations and the elements of the
+  // new user with whole shape.
+  return sum >= xla::ShapeUtil::ElementsIn(operands[0]->shape());
+}
+
+// Collects the peer operations of inst included inst.
+// The operands of each operation in group have the same opcode and they are
+// slices taken from the same indices of bigger tensors with the same
+// dimensions. The corresponding operands of all operations are slices taken
+// from the same bigger tensors.
+// Returns an empty vector if the accumulated size of the operations in group
+// is less than the size of such a bigger tensor. This is a heuristic to
+// ensure that the transformation never causes us to do more elementwise
+// operations
+absl::optional<std::vector<HloInstruction*>> FindPeerElementwiseOperations(
+    const HloInstruction* inst,
+    const std::vector<HloInstruction*>& slice_sources) {
+  std::vector<HloInstruction*> peer_operations;
+  HloInstruction* slice_source0 = slice_sources[0];
+
+  // Traverses the slices which are taken from slice source 0.
+  for (const HloInstruction* operand_slice0 : slice_source0->users()) {
+    // Skips not-slices
+    if (operand_slice0->opcode() != HloOpcode::kSlice) {
+      continue;
+    }
+
+    // The user of slice is a candidate of peer operation.
+    for (HloInstruction* user : operand_slice0->users()) {
+      // The inst's peers should be the same operation and same operand count as
+      // inst.
+      if (user->opcode() != inst->opcode() ||
+          user->operand_count() != inst->operand_count() ||
+          !IsPeerInstruction(user, slice_sources)) {
+        continue;
+      }
+
+      // Found the peer operation.
+      peer_operations.push_back(user);
+    }
+  }
+
+  // Calculates the costs. If cost is more than profit, returns nullopt.
+  return ShouldReplace(slice_sources, peer_operations) ?
+      absl::make_optional(peer_operations) : absl::nullopt;
```

Comment:
my bazel is 0.22.0, if the make_optional is removed, it will cause below error.

error: no match for ternary 'operator?:' (operand types are 'bool', 'std::vector<xla::HloInstruction*>', and 'absl::internal::identity_t<absl::nullopt_t> {aka const absl::nullopt_t}')
       (peer_operations) : absl::nullopt;